apps/emqx_resource/src/emqx_resource_buffer_worker.erl

```diff
@@ -715,17 +714,20 @@ batch_reply_caller_defer_metrics(Id, BatchResult, Batch, QueryOpts) ->
         end,
         Batch
     ),
-    {ShouldAck, PostFns} =
+    {Action, PostFn1} = reply_caller_defer_metrics(Id, hd(Replies), QueryOpts),
+    PostFns =
         lists:foldl(
-            fun(Reply, {_ShouldAck, PostFns}) ->
-                {ShouldAck, PostFn} = reply_caller_defer_metrics(Id, Reply, QueryOpts),
-                {ShouldAck, [PostFn | PostFns]}
+            fun(Reply, PostFns) ->
+                {Action1, PostFn} = reply_caller_defer_metrics(Id, Reply, QueryOpts),
+                %% assert
+                true = (Action1 =:= Action),
```

Comment:
Maybe we don't need this split and assert.  As the result is the same for the whole batch, the verdict should be the same for all replies.